`hiredis-client` on macOS can not find OpenSSL from Homebrew

When installing `hiredis-client` on macOS, it fails with `fatal error: 'openssl/ssl.h' file not found`

This happens when installing with `bundler` and directly with `gem install`

OpenSSL is installed through Homebrew, Ruby is installed using `mise`

`gem install` output:
```
❯ gem install hiredis-client
Building native extensions. This could take a while...
ERROR:  Error installing hiredis-client:
        ERROR: Failed to build gem native extension.

    current directory: /Users/renchap/.local/share/mise/installs/ruby/4.0.6/lib/ruby/gems/4.0.0/gems/hiredis-client-0.30.0/ext/redis_client/hiredis
/Users/renchap/.local/share/mise/installs/ruby/4.0.6/bin/ruby extconf.rb
checking for openssl/ssl.h... yes
cc -std=c99 -pedantic -c -O3 -fPIC  -fvisibility=hidden -Wall -W -Wstrict-prototypes -Wwrite-strings -Wno-missing-field-initializers -g -ggdb alloc.c
cc -std=c99 -pedantic -c -O3 -fPIC  -fvisibility=hidden -Wall -W -Wstrict-prototypes -Wwrite-strings -Wno-missing-field-initializers -g -ggdb net.c
net.c:165:50: warning: unused parameter 'interval' [-Wunused-parameter]
  165 | int redisKeepAlive(redisContext *c, int ttl, int interval) {
      |                                                  ^
1 warning generated.
cc -std=c99 -pedantic -c -O3 -fPIC  -fvisibility=hidden -Wall -W -Wstrict-prototypes -Wwrite-strings -Wno-missing-field-initializers -g -ggdb hiredis.c
cc -std=c99 -pedantic -c -O3 -fPIC  -fvisibility=hidden -Wall -W -Wstrict-prototypes -Wwrite-strings -Wno-missing-field-initializers -g -ggdb sds.c
cc -std=c99 -pedantic -c -O3 -fPIC  -fvisibility=hidden -Wall -W -Wstrict-prototypes -Wwrite-strings -Wno-missing-field-initializers -g -ggdb async.c
cc -std=c99 -pedantic -c -O3 -fPIC  -fvisibility=hidden -Wall -W -Wstrict-prototypes -Wwrite-strings -Wno-missing-field-initializers -g -ggdb read.c
cc -std=c99 -pedantic -c -O3 -fPIC  -fvisibility=hidden -Wall -W -Wstrict-prototypes -Wwrite-strings -Wno-missing-field-initializers -g -ggdb sockcompat.c
ar rcs libhiredis.a alloc.o net.o hiredis.o sds.o async.o read.o sockcompat.o
cc -std=c99 -pedantic -c -O3 -fPIC  -fvisibility=hidden -Wall -W -Wstrict-prototypes -Wwrite-strings -Wno-missing-field-initializers -g -ggdb ssl.c
ssl.c:45:10: fatal error: 'openssl/ssl.h' file not found
   45 | #include <openssl/ssl.h>
      |          ^~~~~~~~~~~~~~~
1 error generated.
make: *** [ssl.o] Error 1
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
        --with-opt-dir
        --without-opt-dir
        --with-opt-include=${opt-dir}/include
        --without-opt-include
        --with-opt-lib=${opt-dir}/lib
        --without-opt-lib
        --with-make-prog
        --without-make-prog
        --srcdir=.
        --curdir
        --ruby=/Users/renchap/.local/share/mise/installs/ruby/4.0.6/bin/$(RUBY_BASE_NAME)
        --with-openssl-dir
        --without-openssl-dir
        --with-openssl-include=${openssl-dir}/include
        --without-openssl-include
        --with-openssl-lib=${openssl-dir}/lib
        --without-openssl-lib
        --with-make-prog
        --without-make-prog
extconf.rb:51:in 'block in HiredisConnectionExtconf#build_hiredis': Building hiredis failed (RuntimeError)
        from extconf.rb:49:in 'Dir.chdir'
        from extconf.rb:49:in 'HiredisConnectionExtconf#build_hiredis'
        from extconf.rb:20:in 'HiredisConnectionExtconf#configure_extension'
        from extconf.rb:12:in 'HiredisConnectionExtconf#configure'
        from extconf.rb:109:in '<main>'

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /Users/renchap/.local/share/mise/installs/ruby/4.0.6/lib/ruby/gems/4.0.0/extensions/arm64-darwin-23/4.0.0-static/hiredis-client-0.30.0/mkmf.log

extconf failed, exit code 1

Gem files will remain installed in /Users/renchap/.local/share/mise/installs/ruby/4.0.6/lib/ruby/gems/4.0.0/gems/hiredis-client-0.30.0 for inspection.
Results logged to /Users/renchap/.local/share/mise/installs/ruby/4.0.6/lib/ruby/gems/4.0.0/extensions/arm64-darwin-23/4.0.0-static/hiredis-client-0.30.0/gem_make.out
```

If I pass the path to the OpenSSL installation directory (`gem install hiredis-client -- --with-openssl-dir=/opt/homebrew/opt/openssl@3`), it builds fine.

As nobody is reporting this I am assuming this is specific to this environment (either OpenSSL from Homebrew or using `mise`), but it makes it non trivial for people to install our codebase with `bundler` if we need to document this workaround. This does not happen for any other gems, so I am wondering if something could be improved for `hiredis-client`?

--- a/hiredis-client/ext/redis_client/hiredis/extconf.rb
+++ b/hiredis-client/ext/redis_client/hiredis/extconf.rb
@@ -59,6 +59,13 @@ class HiredisConnectionExtconf
 
   def configure_openssl(original_env)
     original_env.dup.tap do |env|
+      if pkg = pkg_config("openssl")
+        cflags, *ldflags = pkg
+        env["CFLAGS"] = concat_flags(env["CFLAGS"], cflags)
+        env["SSL_LDFLAGS"] = concat_flags(env["SSL_LDFLAGS"], *ldflags)
+        return env
+      end
+
       config = dir_config("openssl")
       if config.none?
         config = dir_config("opt").map { |c| detect_openssl_dir(c) }

